arangod/Agency/Node.cpp

```diff
@@ -696,13 +697,165 @@ ResultT<std::shared_ptr<Node>> Node::handle<SHIFT>(VPackSlice const& slice) {
+      for (auto const& i : VPackArrayIterator(this->slice())) {
+        if (!i.isEqualString(user.stringRef())) {
+          newValue.add(i);
+        }
+      }
+    }
+    Slice newValueSlice = newValue.slice();
+    if (newValueSlice.length() == 0) {
+      return deleteMe();
+    } else {
+      this->operator=(newValue.slice());
+    }
+    return ResultT<std::shared_ptr<Node>>::success(nullptr);
+  }
+
+  return ResultT<std::shared_ptr<Node>>::error(
+    TRI_ERROR_FAILED, "Read unlock failed");
+
+}
+
+template<>
+ResultT<std::shared_ptr<Node>> Node::handle<WRITE_LOCK>(VPackSlice const& slice) {
+  Slice user = slice.get("by");
+  if (!user.isString()) {
+    return ResultT<std::shared_ptr<Node>>::error(
+      TRI_ERROR_FAILED, std::string("Invalid write unlock: ") + slice.toJson());
+  }
+
+  if (isWriteLockable(user.stringRef())) {
+    this->operator=(user);
+    return ResultT<std::shared_ptr<Node>>::success(nullptr);
+  }
+  return ResultT<std::shared_ptr<Node>>::error(TRI_ERROR_LOCKED);
+
+}
+
+template<>
+ResultT<std::shared_ptr<Node>> Node::handle<WRITE_UNLOCK>(VPackSlice const& slice) {
+  Slice user = slice.get("by");
+  if (!user.isString()) {
+    return ResultT<std::shared_ptr<Node>>::error(
+      TRI_ERROR_FAILED, std::string("Invalid write unlock: ") + slice.toJson());
+  }
+
+  if (isWriteUnlockable(user.stringRef())) {
+    return deleteMe();
+  }
+
+  return ResultT<std::shared_ptr<Node>>::error(
+    TRI_ERROR_FAILED, "Write unlock failed");
+}
+
+bool Node::isReadLockable(const VPackStringRef& by) const {
+  if (!_children.empty()) {
+    return false;
+  }
+  Slice slice = this->slice();
+  // the following states are counted as readLockable
+  // array - when read locked or read lock released
+  // empty object - when the node is created
+  if (slice.isArray()) {
+    // check if `by` is not in the array
+    for (auto const& i : VPackArrayIterator(slice)) {
+      if (!i.isString() || i.isEqualString(VPackStringRef(by.data(), by.length()))) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  return slice.isEmptyObject();
+}
+
+bool Node::isReadUnlockable(const VPackStringRef& by) const {
+  if (!_children.empty()) {
+    return false;
+  }
+  Slice slice = this->slice();
+  // the following states are counted as readUnlockable
+  // array of strings containing the value `by`
+  if (slice.isArray()) {
+    bool valid = false;
+    for (auto const& i : VPackArrayIterator(slice)) {
+      if (!i.isString()) {
+        valid = false;
+        break;
+      }
+      if (i.isEqualString(VPackStringRef(by.data(), by.length()))) {
+        valid = true;
+      }
+    }
+    return valid;
+  }
+  return false;
+}
+
+bool Node::isWriteLockable(const VPackStringRef& by) const {
+  if (!_children.empty()) {
+    return false;
+  }
+  Slice slice = this->slice();
+  // the following states are counted as writeLockable
+  //  empty object - when the node is create
+  return slice.isEmptyObject();
```

Comment:
How about an empty array?

Reply:
An empty array can never happen. If the last readlock is released, the node is deleted and thus will be classified as `emptyObject`.